src/main/java/ch/njol/skript/lang/util/SimpleExpression.java

```diff
@@ -200,8 +200,10 @@ public static <T> boolean check(@Nullable T[] values, Checker<? super T> checker
 	@Nullable
 	@SuppressWarnings("unchecked")
 	public <R> Expression<? extends R> getConvertedExpression(Class<R>... to) {
-		if (CollectionUtils.containsSuperclass(to, getReturnType()))
-			return (Expression<? extends R>) this;
+		for (Class<? extends T> type : this.possibleReturnTypes()) {
+			if (CollectionUtils.containsSuperclass(to, type))
+				return (Expression<? extends R>) this;
+		}
```

Comment:
I'm not sure this is actually something we can do. What happens if the expression doesn't end up returning something that is of the return type assumed here?

To give an example, if we have some Expression with return type Object and possible return types String and Integer (we'll figure out which is actually being returned at runtime), I don't think we can say it's an Expression<String> if it might actually be an Expression<Integer>

Reply:
Removing this breaks arithmetic with variables entirely.
The conversion behaviour for Objects is to cast them, but their return type is still Object, which means arithmetic won't accept them during parse-time. We **have** to make an assumption about the return type in order for parsing to work correctly.

Since the _entire_ fix hinges on this assumption (here and during expression conversion) we'll have to close and abandon fixing this if you don't think it's permissible.

I'm going to attempt something else just to see whether it'll work but there's nothing we can do except assume, or we end up in parsing hell again.